coreneuron/mpi/spdlog_mpi_logger.h

```diff
@@ -0,0 +1,94 @@
+/* Example
+#include <spdlog_mpi_logger.h>
+
+int main(int argc, char *argv[]) {
+    MPI_Init(&argc, &argv);
+    auto logger = std::make_shared<spdlog::mpi_logger>("my_mpi_logger");
+    logger->info("ceci est un test");
+    MPI_Finalize();
+    return 0;
+}
+*/
+
+#pragma once
+
+#include <iostream>
+#include <spdlog/logger.h>
+#include <mpi.h>
+
+namespace spdlog {
+
+class mpi_logger: public logger
+{
+public:
+    // Empty logger
+    explicit mpi_logger(std::string logger_name, int server_rank_ = 0, MPI_Comm comm_ = MPI_COMM_WORLD)
+        : logger(std::move(logger_name))
+        , server_rank(server_rank_)
+    {
+        MPI_Comm_dup(comm_, &comm);
+        MPI_Comm_rank(comm, &mpi_rank);
+        if (mpi_rank == server_rank) {
+            server_thr = std::thread([&](){
+                int stopped = 1;
+                MPI_Finalized(&stopped);
+                while(!stopped) {
```

Comment:
OK, I see how you thought about it. May I suggest an alternate way do this?
The more MPI way of doing parallelism is that all ranks execute essentially the same program. This would mean that in a given piece of code where you'd want to log, all ranks would call `mpi_logger.log(...)` (or something similar). the `log` function could then do two different things:

1. You could go the serialization route: You write a loop over the ranks and serialize with barriers the output, printing at each iteration the log of one rank and protecting it with a barrier. This way you ensure that the ranks don't write over each other, you can then also add fancy things like masks for ranks to print or mute, etc.
2. you could go the `MPI_Gather` route: You first gather (potential) logs from all ranks to rank 0 and print from there. This could be possibly the more efficient way of doing it, but I'm not sure right now.

I'm sure @pramodk has many improvements over my ideas, but I think generally this will be the better approach then the way method here. By the way, the method you use also makes perfectly sense and I've seen it (maybe implemented it in a previous life) in client-server architectures.

Final question for discussion: Would the more MPI way of doing it work with spdlog? I don't see immediately why not, but I haven't thought very hard about this.